src/model_provider/model_provider.h

```diff
@@ -79,6 +80,71 @@ class FileModelProvider : public ModelProvider {
   std::string model_ = {};
 };
 
+
+namespace detail {
+
+inline std::string GenerateAutoserializeCmd(const std::string& module_path,
+                                            const std::string& target_file_path) {
+  std::stringstream cmd;
+
+  cmd << R"py(python3 -c "
+import importlib, sys
+from nvidia.dali._utils.autoserialize import invoke_autoserialize
+spec = importlib.util.spec_from_file_location('autoserialize_mod', ')py" << module_path << R"py(')
+head_module = importlib.util.module_from_spec(spec)
+sys.modules['autoserialize_mod'] = head_module
+spec.loader.exec_module(head_module)
+invoke_autoserialize(head_module, ')py" << target_file_path << R"py(')
+")py";
+
+  return cmd.str();
+}
+
+inline void CallSystemCmd(const std::string& cmd) {
+  auto status = system(cmd.c_str());
+
+  if (WIFEXITED(status) && WEXITSTATUS(status) == 0)
+    // Succeed
+    return;
+
+  // Failed
+  std::stringstream ss;
+  ss << "Failed to call system command. ";
+  if (WIFEXITED(status)) {
+    ss << "Exited with error code: " << WEXITSTATUS(status);
+  } else if (WIFSIGNALED(status)) {
+    ss << "Killed by signal: " << WTERMSIG(status);
+  } else if (WIFSTOPPED(status)) {
+    ss << "Stopped by signal: " << WSTOPSIG(status);
+  } else if (WIFCONTINUED(status)) {
+    ss << "Continued";
+  }
+  throw std::runtime_error(ss.str());
+}
+
+}  // namespace detail
+
+
+class AutoserializeModelProvider : public ModelProvider {
+ public:
+  AutoserializeModelProvider() = default;
+
+  AutoserializeModelProvider(const std::string& module_path, const std::string& target_file) {
+    auto cmd = detail::GenerateAutoserializeCmd(module_path, target_file);
+    detail::CallSystemCmd(cmd);
+    fmp_ = std::make_unique<FileModelProvider>(target_file);
+  }
+
+  const std::string& GetModel() const override {
+    return fmp_->GetModel();
+  }
+
+  ~AutoserializeModelProvider() override = default;
+
+ private:
+  std::unique_ptr<FileModelProvider> fmp_;
```

Comment:
What's the reason for using `unique_ptr` instead of just `FileModelProvider`?

Reply:
Good catch, this doesn't need to be a pointer.